### PR TITLE
API, Core: Multi-Table transactions

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+     * compared to when the transaction started. Loading table data when it is referenced.
+     */
+    SERIALIZABLE_WITH_LOADING_READS,
+
+    /**
+     * All tables participating in the transaction must be in the same state when committing
+     * compared to when the transaction started. Loading table data as of TX start time.
+     */
+    SERIALIZABLE_WITH_FIXED_READS,
+  }
```
Comment (amogh-jahagirdar) [inline api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java:42]: I know this is still draft, but I wanted to verify for multi-table transactions shouldn't the isolation level enforcement be done at the branch level and not just the table level?

For example, for serializable with fixed reads I would expect the validation that the branch is in the same state even if another branch is written to.

```
    CatalogTransaction catalogTx = catalog.startTransaction(isolationLevel);
    Catalog txCatalog = catalogTx.asCatalog();
    catalog.loadTable(table1).newFastAppend().appendFile(FILE_B).toBranch("branch1").commit();
    catalog.loadTable(table2).newFastAppend().appendFile(FILE_C).toBranch("branch2").commit();
```

Let's say the append to table 1 above happens concurrently with an UPDATE/DELETE which matches but the update/delete happens on "branch3", then the multi-table transaction should still succeed.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalogTransaction.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    txByTable().forEach((key, value) -> updatesByTable().put(key, value.pendingUpdates()));
+
+    try {
+      if (IsolationLevel.SERIALIZABLE_WITH_FIXED_READS == isolationLevel()
+          || IsolationLevel.SERIALIZABLE_WITH_LOADING_READS == isolationLevel()) {
+        // make sure tables that are participating in the TX haven't been updated externally
+        for (TableIdentifier affectedTable : updatesByTable().keySet()) {
+          TableMetadata tableMetadata =
+              ((BaseTable) origin().loadTable(affectedTable)).operations().current();
+          if (tableMetadata.lastUpdatedMillis() >= txStartTime().toEpochMilli()) {
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalogTransaction.java:70]: This goes back to my point above, I think checking tableMetadata last updated is too broad when performing the isolation level checks, I think it needs to be performed at the branch level. Unless we're establishing that multi-table transactions can only happen on each table's main branch but that feels too restrictive.

**Review by amogh-jahagirdar — COMMENTED** [review]
This is awesome to see @nastra I'll hold off on reviewing the code itself until it's out of draft. I did have more fundamental questions on behavior, especially with multi-table transactions in the presence of branches, and how the isolation level checks are performed.

**Comment by rdblue** [comment]
@nastra, can you write up a design doc?

File: api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java (hunk 1 above)
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java:42]: it should actually be most likely at the branch level (but I'd have to revisit and think about this). However, I started to work on this before writing to branches was committed, hence it's not being reflected here.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalogTransaction.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalogTransaction.java:70]: same as above, I started working on this before writing to branches was possible, hence I didn't consider this in this initial version

File: api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * Table can change underneath but still have to pass normal table validation. Loading table
+     * data when it's referenced.
+     */
+    SNAPSHOT,
+
+    /**
+     * All tables participating in the transaction must be in the same state when committing
+     * compared to when the transaction started. Loading table data when it is referenced.
+     */
+    SERIALIZABLE_WITH_LOADING_READS,
```
Comment (zinking) [inline api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java:35]: does this mean data operated in the transaction could be seen during the transaction?

Comment (nastra) [reply api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java:35]: `SERIALIZABLE_WITH_LOADING_READS` that you'd see data at the time of when the TX started and not the modified data updated as part of the TX. If you'd want that, then you'd use one of the other isolation levels

Comment (zinking) [reply api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java:35]: `Loading table data as of TX start time.` - `SERIALIZABLE_WITH_FIXED_READS`  isn't it the other way round ?

if these are semantically consistent with DB isolation levels, we can keep the names consistent.

**Comment by nastra** [comment]
Closing this in favor of https://github.com/apache/iceberg/pull/6948

**Pull request closed without merging**